conda_smithy/rattler_linter.py

```diff
@@ -0,0 +1,590 @@
+        lints.append(
+            "The top level meta keys are in an unexpected order. "
+            "Expecting {}.".format(section_order_sorted_str)
+        )
+
+
+def lint_about_contents(about_section, lints):
+    for about_item in ["homepage", "license", "summary"]:
+        # if the section doesn't exist, or is just empty, lint it.
+        if not about_section.get(about_item, ""):
+            lints.append(
+                "The {} item is expected in the about section."
+                "".format(about_item)
+            )
+
+
+def lint_recipe_tests(test_section=dict(), outputs_section=list()):
+    TEST_KEYS = {"script", "python"}
+    lints = []
+    hints = []
+
+    if not any(key in TEST_KEYS for key in test_section):
+        if not outputs_section:
+            lints.append("The recipe must have some tests.")
+        else:
+            has_outputs_test = False
+            no_test_hints = []
+            for section in outputs_section:
+                test_section = section.get("tests", {})
+                if any(key in TEST_KEYS for key in test_section):
+                    has_outputs_test = True
+                else:
+                    no_test_hints.append(
+                        "It looks like the '{}' output doesn't "
+                        "have any tests.".format(section.get("name", "???"))
+                    )
+            if has_outputs_test:
+                hints.extend(no_test_hints)
+            else:
+                lints.append("The recipe must have some tests.")
+
+    return lints, hints
+
+
+def lint_requirements_order(requirements_section: Dict, lints: List):
+    seen_requirements = [
+        k for k in requirements_section if k in REQUIREMENTS_ORDER
+    ]
+    requirements_order_sorted = sorted(
+        seen_requirements, key=REQUIREMENTS_ORDER.index
+    )
+    if seen_requirements != requirements_order_sorted:
+        lints.append(
+            "The `requirements/` sections should be defined "
+            "in the following order: "
+            + ", ".join(REQUIREMENTS_ORDER)
+            + "; instead saw: "
+            + ", ".join(seen_requirements)
+            + "."
+        )
+
+
+def lint_has_recipe_file(about_section, lints):
+    license_file = about_section.get("license_file", None)
+    if not license_file:
+        lints.append("license_file entry is missing, but is required.")
+
+
+def lint_package_name(package_section: Dict, context_section: Dict):
+    package_name = str(package_section.get("name"))
+    context_name = str(context_section.get("name"))
+    actual_name = (
+        package_name
+        if package_name is not None and not package_name.startswith("$")
+        else context_name
+    )
+
+    actual_name = actual_name.strip()
+    if re.match(r"^[a-z0-9_\-.]+$", actual_name) is None:
+        return """Recipe name has invalid characters. only lowercase alpha, numeric, underscores, hyphens and dots allowed"""
+
+
+def run_conda_forge_specific(
```

Comment:
The duplication in this function is especially egregious. Let's fix that before merging as it is a few line refactor I think.

Reply:
done - we now pass rattler_lint to differentiate how to extract necessary fields
let me know wdyt

Reply:
thank you!